### PR TITLE
Snapshot adapter for typed persistence

### PR DESCRIPTION
**Pull request by chbatey**
Refs #27248 

Introduction of a snapshot adapter to aid in the migration from persistent FSM.

Not sure this is worth highlighting in reference docs in addition to what we'll add for the persistent FSM migration guide in #26490

### CODE REVIEW

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/SnapshotAdapter.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+   * Can be used for migrations from different serialized state types.
+   */
+  def fromJournal(from: Any): State
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] class NoOpSnapshotAdapter extends SnapshotAdapter[Any] {
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/SnapshotAdapter.scala]: move to `internal` package?

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedSnapshotAdapterSpec.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  private val conf: Config = ConfigFactory.parseString(s"""
+    akka.persistence.journal.leveldb.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+    akka.persistence.journal.plugin = "akka.persistence.journal.leveldb"
+    akka.persistence.snapshot-store.plugin = "akka.persistence.snapshot-store.local"
+    akka.persistence.snapshot-store.local.dir = "target/typed-persistence-${UUID.randomUUID().toString}"
+  """)
+  case class State(s: String)
+  case class Command(c: String)
+  case class Event(e: String)
+  case class PersistedState(s: String)
```
Comment (patriknw) [inline akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedSnapshotAdapterSpec.scala]: `extends CborSerializable` when my PR has been merged

**Review by patriknw — APPROVED** [review]
LGTM, with one nitpick

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/NoOpSnapshotAdapter.scala [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  override def toJournal(state: Any): Any = state
+  override def fromJournal(from: Any): Any = from
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+object NoOpSnapshotAdapter {
+  val i = new NoOpSnapshotAdapter
```
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/NoOpSnapshotAdapter.scala]: private?

**Review by johanandren — APPROVED** [review]

**Pull request merged**